### ISSUE TITLE
[iOS][2.0.0] Not awake closed app when received voip call.

### ISSUE THREAD

**Issue by william-normann** [react-native-webrtc/react-native-voip-push-notification#48]
We use `voip-push-notification` & `react-native-callkeep` on iOS to awake app from background mode / closed state when received a voip call.

It work fine in background mode. But when app in closed state. Nothing happen, no notifcation, no automation awake app.

We met this issue from version 2.0.0 and react native 0.60.6. V1 and react-native 0.59 work fine.

any ones met the same issue with us? Can we resolve this problem?

**Comment by tuanbm** [comment]
Have you got any workaround for this?

**Comment by zxcpoiu** [comment]
Please read the new notes in the readme to see if there any missed part.

https://github.com/react-native-webrtc/react-native-voip-push-notification/blob/master/README.md#important-note

**Comment by ThanhNghia944** [comment]
@william-normann  pls for me know how to create voip  to test react-native-voip-push-notification

**Comment by ThanhNghia944** [comment]
same issue

**Comment by zxcpoiu** [comment]
Have you find the cause? I know ios 13.3 seems have issues, butsolved after upgrade to ios 13.4

**Comment by MrAlexWeber** [comment]
I'm running into a similar issue. It seems like `VoipPushNotification.wakeupByPush` is `false`, and I'm not sure why.

**Comment by MrAlexWeber** [comment]
> Have you got any workaround for this?

As a workaround you can `import { AppState } from 'react-native'` and then basically just check that `Platform.OS === 'ios' && AppState.currentState === 'background'` instead of checking `wakeupByPush`.

**Comment by tuanbm** [comment]
> > Have you got any workaround for this?
> 
> As a workaround you can `import { AppState } from 'react-native'` and then basically just check that `Platform.OS === 'ios' && AppState.currentState === 'background'` instead of checking `wakeupByPush`.

It's the problem with iOS version < 13, I checked with iOS 12. 
My work around is to send double voip notification (and of cause handle the duplication on client code, for me, I send APN messages with a flag to mark if a notification is a duplication one), so far it's working for me. Btw, ios 3 is working as expected, no need to apply the workaround.

**Comment by MrAlexWeber** [comment]
> It's the problem with iOS version < 13, I checked with iOS 12.
> Btw, ios 13 is working as expected, no need to apply the workaround.

I was having the issue when testing on iOS 13.4

**Comment by vsalle** [comment]
> > It's the problem with iOS version < 13, I checked with iOS 12.
> > Btw, ios 13 is working as expected, no need to apply the workaround.
> 
> I was having the issue when testing on iOS 13.4

Hi, I have the same issue in iOS 13.4.1 i can't wake up app.

**Comment by willnaoosmit** [comment]
did you guys even managed to use it with the app minimized? I'm struggling with it and dont find anything to make it work :(

**Comment by pschill** [comment]
The problem is that the method didReceiveIncomingPushWithPayload in AppDelegate.m doesn't get called when the app is in the background. This results in the system trying to invoke Callkit a few times but fails and it then drops all attempts to start your application on a background push (the console log will just show "Dropped on the floor").

To enable the retries to call this method you need to reinstall your application. 

I'm really new at iOS-development so I have no idea why this method doesn't get called or how you call it from a push.

**Comment by kentanaka000** [comment]
> It's the problem with iOS version < 13, I checked with iOS 12.
> My work around is to send double voip notification (and of cause handle the duplication on client code, for me, I send APN messages with a flag to mark if a notification is a duplication one), so far it's working for me. Btw, ios 13 is working as expected, no need to apply the workaround.

How exactly did you get this workaround to work? I've been trying this and it gives me weird bugs when the call is closed before the duplicate comes in

**Comment by tuanbm** [comment]
> > It's the problem with iOS version < 13, I checked with iOS 12.
> > My work around is to send double voip notification (and of cause handle the duplication on client code, for me, I send APN messages with a flag to mark if a notification is a duplication one), so far it's working for me. Btw, ios 13 is working as expected, no need to apply the workaround.
> 
> How exactly did you get this workaround to work? I've been trying this and it gives me weird bugs when the call is closed before the duplicate comes in

It's a little bit tricky that the duplicated push should be sent just after the 1st one came, like: send the 1st one, delay 1s, then send the 2nd one. 2 notification should contain the same uuid and in the code, you should call the api to show the incoming call UI with the same UUID ( that means 2nd one will be ignored if the UI is still showing because of the 1st one)
Hope this helps.

**Comment by pschill** [comment]
Finally got it to work.

RNCallKeep.setup(options).then(accepted => {
    });

Needs to be setup in index.js. Or else everything just fails silently.

**Comment by pschill** [comment]
Finally "solved" it. I know nothing about iOS-development so my "solution" is probably pretty shitty. The problem is that the OS is waking the application before the javascript code initializes this lib.

I went into node_modules and this project (will have to upload the changes into a local git repo later) and then I changed the method voipRegistration to be static (+ instead of -) and removed 

dispatch_async(dispatch_get_main_queue(), ^{
        [self voipRegistration];
    });
 from RCT_EXPORT_METHOD(registerVoipToken) since this can not be called from the instance anymore.

Then instead of setting up voip from this library, you call it from didFinishLaunchingWithOptions in AppDelegate.m in your own project so that your app has time to set it up before didReceiveIncomingPushWithPayload comes.

So just add this line:
[RNVoipPushNotificationManager voipRegistration];
Somewhere before the return in didFinishLaunchingWithOptions.

**Comment by pschill** [comment]
Finally "solved" it. I know nothing about iOS-development so my "solution" is probably pretty shitty. The problem is that the OS is waking the application before the javascript code initializes this lib.

I went into node_modules and this project (will have to upload the changes into a local git repo later) and then I duplicated the method voipRegistration into a static one as well (+ instead of -).

Then I called it from didFinishLaunchingWithOptions in AppDelegate.m in my own project so that the app had time to set it up before didReceiveIncomingPushWithPayload came.

In conclusion:
1. Duplicate voipRegistration-merthod in RNVoipPushNotificationManager into another static one (+ instead of -)
2. Call [RNVoipPushNotificationManager voipRegistration];
Somewhere before the return in didFinishLaunchingWithOptions in your AppDelegate.m.

**Comment by zxcpoiu** [comment]
please check discussion in #59 and the test the fix in PR #69

**Comment by zxcpoiu** [comment]
Released 3.0.0, this should be fixed. Please check doc and test, especially `didLoadWithEvents` part.
